clients/python-wrapper/lakefs/object.py

```diff
@@ -2,19 +2,380 @@
+            ObjectNotFoundException if repo id, reference id or object path does not exist
+            PermissionException if user is not authorized to perform this operation, or operation is forbidden
+            ServerException for any other errors
+        """
+        if mode not in get_args(WriteModes):
+            raise ValueError(f"invalid write mode: '{mode}'. WriteModes: {WriteModes}")
+
+        content = data
+        if 'x' in mode and self.exists():  # Requires explicit create
+            raise ObjectExistsException
+
+        # TODO: handle streams
+        binary_mode = 'b' in mode
+        if binary_mode and isinstance(data, str):
+            content = data.encode('utf-8')
+        elif not binary_mode and isinstance(data, bytes):
+            content = data.decode('utf-8')
+
+        is_presign = pre_sign if pre_sign is not None else self._client.storage_config.pre_sign_support
+        with api_exception_handler(_io_exception_handler):
+            stats = self._upload_presign(content, content_type, metadata) if is_presign \
+                else self._upload_raw(content, content_type, metadata)
+            self._stats = ObjectStats(**stats.__dict__)
+
+        return self
+
+    def delete(self) -> None:
+        """
+        Delete object from lakeFS
+            ObjectNotFoundException if repo id, reference id or object path does not exist
+            PermissionException if user is not authorized to perform this operation, or operation is forbidden
+            ServerException for any other errors
+        """
+        with api_exception_handler(_io_exception_handler):
+            self._client.sdk_client.objects_api.delete_object(self._repo_id, self._ref_id, self._path)
+            self._stats = None
+
+    @staticmethod
+    def _extract_etag_from_response(headers) -> str:
+        # prefer Content-MD5 if exists
+        content_md5 = headers.get("Content-MD5")
+        if content_md5 is not None and len(content_md5) > 0:
+            try:  # decode base64, return as hex
+                decode_md5 = base64.b64decode(content_md5)
+                return binascii.hexlify(decode_md5).decode("utf-8")
+            except binascii.Error:
+                pass
+
+        # fallback to ETag
+        etag = headers.get("ETag", "").strip(' "')
+        return etag
+
+    def _upload_raw(self, content: UploadContentType, content_type: Optional[str] = None,
+                    metadata: Optional[dict[str, str]] = None) -> ObjectStats:
+        """
+        Use raw upload API call to bypass validation of content parameter
+        """
+        headers = {
+            "Accept": "application/json",
+            "Content-Type": content_type if content_type is not None else "application/octet-stream"
+        }
+
+        # Create user metadata headers
+        if metadata is not None:
+            for k, v in metadata.items():
+                headers[_LAKEFS_METADATA_PREFIX + k] = v
```

Comment:
validate that the underlaying call_api code validate the header encoding - we like to get an error or encode the right value as part of the header name.
Example posting key='a:key' will probably fail.

Reply:
Can you explain why this example will fail?

Reply:
Verified - the httplib throws a ValueError exception